boa/vm/py_evm.py

```diff
@@ -0,0 +1,597 @@
+# every time one is created. the reasoning is that it would seem
+# confusing to have registered precompiles not persist across envs -
+# if somebody registers a precompile, presumably they want it to work
+# on all envs.
+_precompiles = {}
+
+
+def register_precompile(*args, **kwargs):
+    warnings.warn(
+        "register_recompile has been renamed to register_raw_precompile!", stacklevel=2
+    )
+
+
+def register_raw_precompile(address, fn, force=False):
+    global _precompiles
+    address = Address(address)
+    if address in _precompiles and not force:
+        raise ValueError(f"Already registered: {address}")
+    _precompiles[address.canonical_address] = fn
+
+
+def deregister_raw_precompile(address, force=True):
+    address = Address(address).canonical_address
+    if address not in _precompiles and not force:
+        raise ValueError("Not registered: {address}")
+    _precompiles.pop(address, None)
+
+
+def console_log(computation):
+    msgdata = computation.msg.data_as_bytes
+    schema, payload = abi_decode("(string,bytes)", msgdata[4:])
+    data = abi_decode(schema, payload)
+    print(*data, file=sys.stderr)
+    return computation
+
+
+CONSOLE_ADDRESS = bytes.fromhex("000000000000000000636F6E736F6C652E6C6F67")
+
+register_raw_precompile(CONSOLE_ADDRESS, console_log)
+
+
+# a code stream which keeps a trace of opcodes it has executed
+class TracingCodeStream(CodeStream):
+    __slots__ = [
+        "_length_cache",
+        "_fake_codesize",
+        "_raw_code_bytes",
+        "invalid_positions",
+        "valid_positions",
+        "program_counter",
+    ]
+
+    def __init__(self, *args, start_pc=0, fake_codesize=None, contract=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._trace = []  # trace of opcodes that were run
+        self.program_counter = start_pc  # configurable start PC
+        self._fake_codesize = fake_codesize  # what CODESIZE returns
+
+    def __iter__(self) -> Iterator[int]:
+        # upstream says: "a very performance-sensitive method"
+        # note: not clear to me that len(raw_code_bytes) is a hotspot
+        while self.program_counter < self._length_cache:
+            opcode = self._raw_code_bytes[self.program_counter]
+
+            self._trace.append(self.program_counter)
+            self.program_counter += 1
+            yield opcode
+
+        yield STOP
+
+    def __len__(self):
+        if self._fake_codesize is not None:
+            return self._fake_codesize
+        return self._length_cache
+
+
+# ### section: sha3 preimage tracing
+def _stackitem_to_int(value):
+    if isinstance(value, tuple):
+        return to_int(value[1])  # how py-evm<=0.8.0b1 stores stuff on stack
+    else:
+        return to_int(value)
+
+
+def _stackitem_to_bytes(value):
+    if isinstance(value, tuple):
+        return to_bytes(value[1])  # how py-evm<=0.8.0b1 stores stuff on stack
+    else:
+        return to_bytes(value)
+
+
+class Sha3PreimageTracer:
```

Comment:
i think these tracer classes should stay in Env tbh

Reply:
they do use py-evm internals so we can just leave a note to make them generic (operate against any evm implementation) in the future